tools/ld-chroma-decoder/ycbcr.cpp

```diff
@@ -0,0 +1,98 @@
+/************************************************************************
+
+    ycbcr.cpp
+
+    ld-chroma-decoder - Colourisation filter for ld-decode
+    Copyright (C) 2018 Chad Page
+    Copyright (C) 2018-2019 Simon Inns
+    Copyright (C) 2021 Phillip Blucas
+
+    This file is part of ld-decode-tools.
+
+    ld-chroma-decoder is free software: you can redistribute it and/or
+    modify it under the terms of the GNU General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+************************************************************************/
+
+#include "ycbcr.h"
+
+
+YCbCr::YCbCr(double _whiteIreLevel, double _blackIreLevel, bool _whitePoint75, double _chromaGain)
+    : whiteIreLevel(_whiteIreLevel), blackIreLevel(_blackIreLevel), whitePoint75(_whitePoint75),
+      chromaGain(_chromaGain)
+{
+}
+
+void YCbCr::convertLine(const YIQ *begin, const YIQ *end, quint16 *outY, quint16 *outCb, quint16 *outCr)
+{
+    // Factors to scale Y according to the black to white interval
+    // (i.e. make the black level 16*235 and the white level 235*256)
+    double yBlackLevel = blackIreLevel;
+    double yScale = 219.0 * 257.0 / (whiteIreLevel - blackIreLevel);
+
+    // Compute I & Q scaling factors
+    const double iqScale = chromaGain / (whiteIreLevel - blackIreLevel);
+    const double cbScale = 112*256 / (ONE_MINUS_Kb * kB); // Poynton, Eq 25.5 & 28.1
+    const double crScale = 112*256 / (ONE_MINUS_Kr * kR);
+
+    if (whitePoint75) {
+        // NTSC uses a 75% white point; so here we scale the result by
+        // 25% (making 100 IRE 25% over the maximum allowed white point).
+        // This doesn't affect the chroma scaling.
+        yScale *= 125.0 / 100.0;
+    }
+
+    for (const YIQ *yiq = begin; yiq < end; yiq++) {
+        double y = yiq->y;
+        double i = yiq->i;
+        double q = yiq->q;
+
+if (0) {
```

Comment:
Dead code - remove.

Reply:
Whoops.  I knew I'd forget to remove that.  Be comforted that 20 IRE +Q and -I vectors do decode correctly. :)  +Q has negative green and -I has negative red when you decode to RGB, so it's a good test case.